### PR TITLE
[W5OC] Skydev's challenge entry ("conservative" edition)

### PR DESCRIPTION
**Pull request by Skydev0h**
This is a "conservative" edition of my contest entry.

The conservativeness of it is that it does not involve modification of the TVM Assembler (`Asm.fif`), but by it, it loses a considerable amount of possible optimizations, since this change very well synergizes with other optimizations (without them such extent would not be possible because of code size constraints), and I'd like to invite you to [take a look at the "radical" edition here:](https://github.com/tonkeeper/w5/pull/4) #4.

You can take a look at all the optimizations with detailed explanations of each optimization and commit at the [Improvements log](https://github.com/Skydev0h/w5/blob/main/Improvements.rst) document (that is also now linked from the [README](https://github.com/Skydev0h/w5) page, that also now has a [special section about the contest](https://github.com/Skydev0h/w5#warning-contest-note)).

Also, an [undoubtable GitHub activity history](https://github.com/Skydev0h/w5p/activity) is provided as a time precedence guarantee that the optimizations are original and are not copy-cated.

### CODE REVIEW

**Comment by Skydev0h** [comment]
It seems that removing difference automatically closes the PR, so I will have to force-push the changed code over the dummy commit 😓

**Comment by Skydev0h** [comment]
The code has been published!

File: Improvements.rst [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+As an example, here is a comparison of used gas in main ("conservative") and entrypoint ("radical") branches:
+
++-----------------+----------+----------------+
+| Test case       | main gas | entrypoint gas |
++=================+==========+================+
+| External        | 2699     | **2707**       |
++-----------------+----------+----------------+
+| Internal        | 3165     | **2963**       |
++-----------------+----------+----------------+
+| Extension       | 1828     | **1626**       |
```
Comment (siandreev) [inline Improvements.rst:131]: the correct values are `| External        | 1810     | **1608**       |`

File: contracts/failed/wallet_v5_asm_cond_multicheck.fc
Comment (siandreev) [inline contracts/failed/wallet_v5_asm_cond_multicheck.fc:1]: Why is this file needed?

File: contracts/failed/wallet_v5_cache_1ext_broken.fc
Comment (siandreev) [inline contracts/failed/wallet_v5_cache_1ext_broken.fc:1]: Why is this file needed?

File: contracts/failed/wallet_v5_defrag.fc
Comment (siandreev) [inline contracts/failed/wallet_v5_defrag.fc:1]: Why is this file needed?

File: contracts/failed/wallet_v5_rewrite_signed.fc
Comment (siandreev) [inline contracts/failed/wallet_v5_rewrite_signed.fc:1]: Why is this file needed?

File: contracts/wallet_v5.fc
Comment (siandreev) [inline contracts/wallet_v5.fc:1]: It would be a good idea to remove all unused function declarations

File: Improvements.rst [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
++----------------------------------------------------------------+------+------+------+-------+------+-------+-------+-------+-------+--------+
+| Use SDBEGINSQ to check internal message prefixes               | 2710 | 3283 | 1736 | 7729  | 2521 | 24.6% | 59237 | 64090 | 33578 | 156905 |
++----------------------------------------------------------------+------+------+------+-------+------+-------+-------+-------+-------+--------+
+| Backport some optimizations from EP and coalesce code          | 2699 | 3165 | 1828 | 7692  | 2558 | 25.0% | 59108 | 63031 | 34141 | 156280 |
++----------------------------------------------------------------+------+------+------+-------+------+-------+-------+-------+-------+--------+
+| Optimized instructions order for extension and fix args        | 2699 | 3165 | 1810 | 7674  | 2576 | 25.1% | 59108 | 63031 | 34033 | 156172 |
++----------------------------------------------------------------+------+------+------+-------+------+-------+-------+-------+-------+--------+
+| *Reminder and origin point: INITIAL*                           | 3235 | 4210 | 2760 | 10250 | 0    | 0.00% | 64038 | 71163 | 38866 | 174067 |
```
Comment (siandreev) [inline Improvements.rst:116]: Wrong initial total, should be 10205 instead of 10250

File: contracts/failed/wallet_v5_asm_cond_multicheck.fc
Comment (Skydev0h) [reply contracts/failed/wallet_v5_asm_cond_multicheck.fc:1]: This was recorded as an example of failed optimization, that, possibly, might have been used in the future, because if conditions do change due to some other optimizations, it would actually work.
But since the contest has ended and no more logic changes are done, it can be safe to delete these files.

File: contracts/failed/wallet_v5_cache_1ext_broken.fc
Comment (Skydev0h) [reply contracts/failed/wallet_v5_cache_1ext_broken.fc:1]: This was recorded as an example of failed optimization, that, possibly, might have been used in the future, because if conditions do change due to some other optimizations, it would actually work.
But since the contest has ended and no more logic changes are done, it can be safe to delete these files.

File: contracts/failed/wallet_v5_defrag.fc
Comment (Skydev0h) [reply contracts/failed/wallet_v5_defrag.fc:1]: This was recorded as an example of failed optimization, that, possibly, might have been used in the future, because if conditions do change due to some other optimizations, it would actually work.
But since the contest has ended and no more logic changes are done, it can be safe to delete these files.

File: contracts/failed/wallet_v5_rewrite_signed.fc
Comment (Skydev0h) [reply contracts/failed/wallet_v5_rewrite_signed.fc:1]: This was recorded as an example of failed optimization, that, possibly, might have been used in the future, because if conditions do change due to some other optimizations, it might actually work somehow.
But since the contest has ended and no more logic changes are done, it can be safe to delete these files.

File: Improvements.rst (hunk 2 above)
Comment (Skydev0h) [reply Improvements.rst:116]: Well, that's a difficult one, I see, but that will require recalculating of all the percentages and differences. Will work on it.

File: contracts/wallet_v5.fc
Comment (Skydev0h) [reply contracts/wallet_v5.fc:1]: That can be done, until the last moment I was not sure which functions might be used in future.

File: Improvements.rst (hunk 1 above)
Comment (Skydev0h) [reply Improvements.rst:131]: That might be quite off, because the comparation was done at a specific point in time and was not updated later.

Comment (Skydev0h) [reply Improvements.rst]: Fixed

File: contracts/failed/wallet_v5_cache_1ext_broken.fc
Comment (Skydev0h) [reply contracts/failed/wallet_v5_cache_1ext_broken.fc:1]: Unneccessary file removed

File: contracts/failed/wallet_v5_defrag.fc
Comment (Skydev0h) [reply contracts/failed/wallet_v5_defrag.fc:1]: Unneccessary file removed

File: contracts/wallet_v5.fc
Comment (Skydev0h) [reply contracts/wallet_v5.fc:1]: Unneccessary file removed

File: Improvements.rst (hunk 2 above)
Comment (Skydev0h) [reply Improvements.rst]: Fixed and recalculated everything

File: contracts/failed/wallet_v5_rewrite_signed.fc
Comment (Skydev0h) [reply contracts/failed/wallet_v5_rewrite_signed.fc:1]: Unneccessary file removed

File: contracts/failed/wallet_v5_asm_cond_multicheck.fc
Comment (Skydev0h) [reply contracts/failed/wallet_v5_asm_cond_multicheck.fc:1]: Unneccessary file removed

File: contracts/wallet_v5.fc
Comment (Skydev0h) [reply contracts/wallet_v5.fc:1]: Removed all unused functions, commented-out old code and optimized comments

**Comment by Skydev0h** [comment]
@siandreev fixed the issues mentioned in the review comments

**Comment by Nathdogg777** [comment]
GHE-EASY

**Review by Scutua — APPROVED** [review]

**Review by johnny5OO — APPROVED** [review]

File: contracts/wallet_v5.fc
Comment (mehransabzevari2211-oss) [reply contracts/wallet_v5.fc:1]: Iwould be grateful if you could donate one or more tons to help buy medicine for the cancer patients' families. I would like to thank each and every one of you for your good health.

UQAyLOXJwNMvUZjhUKG54LryZsHLUAD8yz0KadIsrKqRa2yH

**Review by johnny5OO — APPROVED** [review]
